#3: Image Component Enhanced

BeeMargarida opened the pull request:
| - | - |
| --- | --- |
| Issue | Closes #3 |
| Dependencies | https://github.com/ripe-tech/ripe-sdk/pull/192 |
| Decisions | Image component with new props (crop, format) and initials logic (initialsBuilder, initialsGroup, showInitials and state). The state contains information about the initials and engraving. |
| Animated GIF | <img width="492" alt="image-personalization" src="https://user-images.githubusercontent.com/25725586/89039270-aa2e9e80-d339-11ea-9995-a41512d5cca2.png">![image_enh](https://user-images.githubusercontent.com/25725586/89039261-a3a02700-d339-11ea-81d1-15952f577e00.gif)|

Commit:
More props and docs

Commit:
Showing state initials when first initialized and Docs

Commit:
Fix comment

Commit:
Use partsData

Commit:
Fix parts

Commit:
Update

Commit:
Refactor

Commit:
Fix documentation

gcandal commented on README.md:
```diff
@@ -181,9 +183,44 @@ The customization of the model can also be provided, with the prop parts:
+
+It is also possible to provide a `initialsBuilder` function, that allows for a more custom logic in the translation from initials and engraving into initials and profiles.
+
+There can be more than one image using the same instance of Ripe SDK:
```
Maybe this should be under a different section, right now it's inside `Image` which I don't think it's what you intended right?

BeeMargarida replied:
@gcandal The part with the multiple SDKs, or the one related to the `initialsBuilder`?

BeeMargarida replied:
Both of them are related to the image

Commit:
Merge branch 'master' into t-ms/3-image_enhanced

gcandal commented:
@BeeMargarida it seems that changing `crop` in the Storybook doesn't produce any changes. Can you please take a look at what's going on?

gcandal commented:
Also, I guess there could be unit tests for this: instantiate an image, change stuff in the props, check that the image URL changes accordingly.

BeeMargarida commented:
> Also, I guess there could be unit tests for this: instantiate an image, change stuff in the props, check that the image URL changes accordingly.

@gcandal Regarding the testing, its a bit more complicated than that. The setup of Ripe SDK is async, however it happens when the component is mounted, so there is no way in the test for waiting for the configuration to end. Its also very complicated by mocking the SDK because of the bindImage method, which must return an Image that implements Observable methods (bind, unbind, etc), which would involve an extensive mock

Commit:
Code simplification

Waiting for config to be called before updating the image with the state. Removed update after updating props, the error was due to storybook.

Commit:
Fix

BeeMargarida commented:
@gcandal Regarding the tests, even by using stubs and mocks, it is not possible. When mounting the component during the tests, the mounting returns before the async methods finish. It makes it impossible to tests behaviour related to the image, which only becomes available after the ripe sdk configuration ends.

gcandal commented on vue/components/organisms/ripe-image/ripe-image.vue:
```diff
@@ -248,7 +234,8 @@ export const RipeImage = {
                     version: this.version,
-                    parts: this.partsData
+                    parts: this.parts,
+                    safe: true
```
💪
